src/services/linode.rs

```diff
@@ -0,0 +1,289 @@
+use std::net::IpAddr;
+
+use serde_derive::{Deserialize, Serialize};
+
+use crate::http::{Error, Request, Response};
+use crate::util::FixedVec;
+
+use super::{one_or_more_string, DdnsService, DdnsUpdateError};
+
+type RecordId = u64;
+type DomainId = u64;
+
+#[derive(Deserialize, Serialize, Clone, Debug, PartialEq, Eq)]
+pub struct Config {
+    token: Box<str>,
+
+    #[serde(deserialize_with = "one_or_more_string")]
+    domains: Vec<Box<str>>,
+
+    /// The time to live expressed in seconds.
+    ///
+    /// Values that are not multiples of 300 will be rounded to the nearest
+    /// multiple by the Linode API.
+    /// See: https://www.linode.com/docs/api/domains/#domain-record-update__request-body-schema
+    ttl: u32,
+}
+
+pub struct Service {
+    config: Config,
+    cached_records: Vec<Record>,
+}
+
+#[derive(Debug, Clone)]
+struct Domain {
+    id: DomainId,
+
+    name: Box<str>,
+}
+
+#[derive(Debug)]
+struct Record {
+    /// Linode uses a master domain (example.com) and encodes different
+    /// records inside it.
+
+    /// The ID of the record, e.g. ID of sub.example.com
+    id: RecordId,
+
+    /// The domain associated to the record.
+    domain_id: DomainId,
+
+    /// The actual name of the record.
+    name: Box<str>,
+
+    kind: RecordKind,
+}
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+enum RecordKind {
+    A,
+    Aaaa,
+}
+
+impl From<Config> for Service {
+    fn from(config: Config) -> Self {
+        let mut config = config;
+        config.token = (String::from("Bearer ") + &config.token).into();
+        Self {
+            config,
+            cached_records: Vec::new(),
+        }
+    }
+}
+
+impl Service {
+    fn parse_error(&self, response: Response) -> Result<(Box<str>, Box<str>), String> {
+        let resp_json = response
+            .into_json::<serde_json::Value>()
+            .map_err(|e| String::from("unable to parse response as JSON:") + &e.to_string())?;
+
+        let errors = resp_json
+            .get("errors")
+            .ok_or_else(|| String::from("expected map"))?;
+
+        let error = errors
+            .get(0)
+            .ok_or_else(|| String::from("expected array"))?;
+
+        let field = error
+            .get("field")
+            .and_then(|m| m.as_str())
+            .or(Some(""))
+            .unwrap()
```

Comment:
minor nitpick: it seems like you can replace `.or(Some("")).unwrap()` with `.unwrap_or("")`

Reply:
Fixed that, but while I'm at it I also have a question: the `field` in the JSON response is the field of the associated JSON request that caused the error, but if no specific field caused an error then it is omitted in the response.

Since I cannot seem to distinguish between a parsing failure or the JSON not having the field, I use the empty string as the default. Should I return an `Option` instead with `None`?
Consider that currently in the `Linode` error type it is unused, since the formatted error message would be weird with an empty string.
Should I make two types of Linode errors? Like a `LinodeGeneric` when no `field` is provided and a `LinodeField` for the other case?

Reply:
It reads to me that `field` is always filled with something, when it is provided in the response. So it's fine to use an empty string to represent the response JSON not having the field, right?

With that in mind I think we can keep `parse_error` the same, but modify `parse_and_check_response()` so that instead of returning `Err(DdnsUpdateError::Linode(reason.into()))?` it does this:

```rust
let error_message = if !field.is_empty() {
    format!("{} (field = {})", reason, field)
} else {
    reason.into()
};
Err(DdnsUpdateError::Linode(error_message))?
```

Reply:
Yeah that seems way simpler, thanks. Maybe I'll also add a comment.